keps/sig-api-machinery/20190226-network-proxy.md

```diff
@@ -0,0 +1,618 @@
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+The plan is to build a system which allows us to control where the traffic
+coming out of the Kube API Server is sent in an extensible manner. We will do
+this by adding a network proxy system. The KAS can be configured to send traffic
+(or not) to one or more of the proxies. These proxies can then be used as
+extensibility points for that traffic. In addition they will allow for control
+of the network.
+
+## Motivation
+
+Everyone [hates the SSH tunnels](https://github.com/kubernetes/kubernetes/issues/54076). They complicate kube-apiserver code a fair amount and only one cloud provider ever bothered to implement them.
+After a generous year of deprecation time, they will be removed in an upcoming release.
+
+In retrospect, having an explicit level of indirection that separates user-initiated network traffic from API server-initiated traffic seems to be a useful concept.
+This gives Cloud providers the ability to control how API server to pod, node and service connectivity is implemented.
+For example, a Cloud provider may choose to run their API server (control network) and the cluster nodes (cluster network) on isolated private IP networks with overlapping
+IP addresses and require custom communications implementation to reach the endpoints.
+Such a layer would also enable audit logging of requests made by the API server and allows for more validation of outgoing API server connections.
+It is worth noting that the majority of this traffic is likely to be encrypted within our tunnel. So the level of audit capability added here may be limited.
+Structuring the API server in this way would also be a forcing function for keeping architectural layering violations out of apiserver.
+This separation of networks can be used to protect against security concerns such as [Security Impact of Kubernetes API server external IP address proxying](https://groups.google.com/d/msg/kubernetes-security-announce/tyd-MVR-tY4/tyREP9-qAwAJ).
+
+### Goals
+
+Allow us to delete the SSH Tunnel/Node Dialer code from the Kube API
+Server.
+Provide a mechanism for fixing
+https://groups.google.com/d/msg/kubernetes-security-announce/tyd-MVR-tY4/tyREP9-qAwAJ.
+Allow isolation of the Control network from the Cluster network.
+
+### Non-Goals
+
+Build a general purpose Proxy which does everything.
+The proxy provided can be used as a guide for users to build their own custom
```

Comment:
I wasn't clear whether "can be used as a guide" was a separate non-goal, or was modifying the "one proxy to rule them all statement".  I suspect you can just remove the "guide" statement.

Reply:
Take a look if the edit makes it clearer.